java/employee-scheduling/src/main/java/org/acme/employeescheduling/solver/EmployeeSchedulingConstraintProvider.java

```diff
@@ -75,49 +74,44 @@ Constraint atLeast10HoursBetweenTwoShifts(ConstraintFactory constraintFactory) {
     }
 
     Constraint oneShiftPerDay(ConstraintFactory constraintFactory) {
-        return constraintFactory.forEachUniquePair(Shift.class, Joiners.equal(Shift::getEmployee),
-                        Joiners.equal(shift -> shift.getStart().toLocalDate()))
+        return constraintFactory.forEachUniquePair(Shift.class, equal(Shift::getEmployee),
+                equal(shift -> shift.getStart().toLocalDate()))
                 .penalize(HardSoftScore.ONE_HARD)
                 .asConstraint("Max one shift per day");
     }
 
     Constraint unavailableEmployee(ConstraintFactory constraintFactory) {
         return constraintFactory.forEach(Shift.class)
-                .filter(shift -> {
-                    Set<LocalDate> unavailableDates = shift.getEmployee().getUnavailableDates();
-                    return unavailableDates.contains(shift.getStart().toLocalDate())
-                        // The contains() check is ignored for a shift ends at midnight (00:00:00).
-                        || (shift.getEnd().isAfter(shift.getStart().toLocalDate().plusDays(1).atStartOfDay())
-                                && unavailableDates.contains(shift.getEnd().toLocalDate()));
-                })
-                .penalize(HardSoftScore.ONE_HARD, EmployeeSchedulingConstraintProvider::getShiftDurationInMinutes)
+                .join(Employee.class, equal(Shift::getEmployee, Function.identity()))
+                .flattenLast(Employee::getUnavailableDates)
+                .filter((shift, unavailableDate) -> shift.getStart().toLocalDate().equals(unavailableDate)
+                        || shift.getEnd().toLocalDate().equals(unavailableDate))
```

Comment:
The filter body should be extracted to a method on `Shift` and reused; it's non-trivial logic which is repeated in 3 constraints.